### PR TITLE
Initial DB refactor.

### PR DESCRIPTION
**Pull request by Alphasite**
Let me know what you think.

### CODE REVIEW

File: src/main/java/postit/server/controller/DatabaseController.java [hunk 1]
```diff
@@ -16,618 +16,374 @@
...
+    public DatabaseController(Database database) {
+        this.database = database;
+    }
+
+    public Account getAccount(String username) {
+        ResultSet rset = null;
+        Account account = null;
+
+        try {
+            Connection connection = database.connect(); PreparedStatement statement = connection.prepareStatement(getAccountSQL);
```
Comment (nw265) [inline src/main/java/postit/server/controller/DatabaseController.java]: :(((( separate lines?

**Comment by nw265** [comment]
I think other than tiny style things it looks good. You essentially copied out sql statements and refactored db connections into one single file correct?

**Comment by Alphasite** [comment]
I changed the database interface slightly, so now each database implements the database interface and gets passed around directly. It made it possible to have a set only database which wouldn't be included in the run binary.

As for the line thing, yeah. I'm not sure how to handle the lines properly. I really don't like how SQL embeds in Java, it just ends up looking like a mess: http://imgur.com/a/bibzO

**Comment by Alphasite** [comment]
If you want to make the change, here is the edited code:
```
private static final String getAccountSQL = "SELECT * FROM " + ACCOUNT + " WHERE `user_name`=?;";

    private static final String addAccountSQL = "INSERT INTO " + ACCOUNT +
            " (`user_name`, `pwd_key`, `email`, `first_name`, `last_name`) VALUES (?,?,?,?,?);";

    private static final String updateAccountSQL = "UPDATE "+ACCOUNT +
            " SET `user_name`=?, `pwd_key`=?, `email`=?, `first_name`=?, `last_name`=? WHERE `user_name`=?;";

    private static final String removeAccountSQL = "DELETE FROM " + ACCOUNT + " WHERE `user_name`=?;";

    private static final String getDirectorySQL = "SELECT * FROM " + DIRECTORY + " WHERE `user_name`=?;";

    private static final String addDirectorySQL = "INSERT INTO " + DIRECTORY + " (`user_name`, `own_path`) VALUES (?,?);";

    private static final String removeDirectorySQL = "DELETE FROM " + DIRECTORY + " WHERE `user_name`=?;";

    private static final String addDirectoryEntrySQL = "INSERT INTO " + DIRECTORY_ENTRY +
            " (`directory_id`, `name`, `encryption_key`) VALUES (?,?,?);";

    private static final String updateDirectoryEntrySQL = "UPDATE " + DIRECTORY_ENTRY + " " +
            "SET `name`=?, `encryption_key`=? WHERE `directory_entry_id`=?;";

    private static final String getDirectoryEntrySQL = "SELECT * FROM " + DIRECTORY_ENTRY + " WHERE `directory_entry_id`=?;";

    private static final String getDirectoryEntryWithNameSQL = "SELECT * FROM " + DIRECTORY_ENTRY +
            " WHERE `directory_id`=? AND `name`=?;";

    private static final String getDirectoryEntriesSQL = "SELECT * FROM " + DIRECTORY_ENTRY + " WHERE `directory_id`=?;";

    private static final String removeDirectoryEntrySQL = "DELETE FROM " + DIRECTORY_ENTRY + " WHERE `directory_entry_id`=?;";

    private static final String addKeychainSQL = "INSERT INTO " + KEYCHAIN +
            " (`directory_entry_id`, `password`, `metadata`) VALUES (?,?,?);";

    private static final String updateKeychainSQL = "UPDATE " + KEYCHAIN +
            " SET `password`=?, `metadata`=? WHERE `directory_entry_id`=?;";

    private static final String getKeychainSQL = "SELECT * FROM " + KEYCHAIN + " WHERE `directory_entry_id`=?;";

    private static final String removeKeychainSQL = "DELETE FROM " + KEYCHAIN + " WHERE `directory_entry_id`=?;";
```

**Comment by Alphasite** [comment]
```
    private static final String getAccountSQL = "SELECT * FROM " + ACCOUNT + " WHERE `user_name`=?;";
    private static final String addAccountSQL = "INSERT INTO " + ACCOUNT + " (`user_name`, `pwd_key`, `email`, `first_name`, `last_name`) VALUES (?,?,?,?,?);";
    private static final String updateAccountSQL = "UPDATE "+ACCOUNT + " SET `user_name`=?, `pwd_key`=?, `email`=?, `first_name`=?, `last_name`=? WHERE `user_name`=?;";
    private static final String removeAccountSQL = "DELETE FROM " + ACCOUNT + " WHERE `user_name`=?;";
    private static final String getDirectorySQL = "SELECT * FROM " + DIRECTORY + " WHERE `user_name`=?;";
    private static final String addDirectorySQL = "INSERT INTO " + DIRECTORY + " (`user_name`, `own_path`) VALUES (?,?);";
    private static final String removeDirectorySQL = "DELETE FROM " + DIRECTORY + " WHERE `user_name`=?;";
    private static final String addDirectoryEntrySQL = "INSERT INTO " + DIRECTORY_ENTRY + " (`directory_id`, `name`, `encryption_key`) VALUES (?,?,?);";
    private static final String updateDirectoryEntrySQL = "UPDATE " + DIRECTORY_ENTRY + " " + "SET `name`=?, `encryption_key`=? WHERE `directory_entry_id`=?;";
    private static final String getDirectoryEntrySQL = "SELECT * FROM " + DIRECTORY_ENTRY + " WHERE `directory_entry_id`=?;";
    private static final String getDirectoryEntryWithNameSQL = "SELECT * FROM " + DIRECTORY_ENTRY + " WHERE `directory_id`=? AND `name`=?;";
    private static final String getDirectoryEntriesSQL = "SELECT * FROM " + DIRECTORY_ENTRY + " WHERE `directory_id`=?;";
    private static final String removeDirectoryEntrySQL = "DELETE FROM " + DIRECTORY_ENTRY + " WHERE `directory_entry_id`=?;";
    private static final String addKeychainSQL = "INSERT INTO " + KEYCHAIN + " (`directory_entry_id`, `password`, `metadata`) VALUES (?,?,?);";
    private static final String updateKeychainSQL = "UPDATE " + KEYCHAIN + " SET `password`=?, `metadata`=? WHERE `directory_entry_id`=?;";
    private static final String getKeychainSQL = "SELECT * FROM " + KEYCHAIN + " WHERE `directory_entry_id`=?;";
    private static final String removeKeychainSQL = "DELETE FROM " + KEYCHAIN + " WHERE `directory_entry_id`=?;";
```

**Comment by Alphasite** [comment]
Uh. Oops, wrong branch.

**Pull request merged**